### ISSUE TITLE
Allow rolling updates to ShardedDaemonProcess

### ISSUE THREAD

**Issue by negokaz** [akka/akka#29250]
We need to update the applicaion two times when we add a process of ShardedDaemonProcess. Otherwise, `behaviorFactory` can fail to produce process behavior. Because not all nodes know the behavior of the new process during a rooling update.

For example:

- There are three nodes NODE-A, NODE-B and NODE-C
- Those nodes started ShardedDaemonProcess as follows

```scala
val tags = Vector("tag-1", "tag-2", "tag-3")
val numberOfProcess = tags.size
val behaviorFactory = id => TagProcessor(tags(id))
ShardedDaemonProcess(system).init("TagProcessors", numberOfProcess, behaviorFactory)
```

- Changed the application to add "tag-4" to `tags`
- When NODE-A was updated, NODE-B and NODE-C doesn't know "tag-4"
- `behaviorFactory` on NODE-B and NODE-C fail to produce the behavior

To get around this problem, we need to update `behaviorFactory` on all nodes before update `numberOfProcess`. 

Another problem is that removing tags other than the last element in tags intorduces the same type of process can be started on multiple nodes. For exmple, if "tags-1" was removed on NODE-1, NODE-1 can start "tag-2" with ID 0 even if "tag-2" with ID 1 is running on other nodes. 

To resolve those problems, I propose to change the ShardedDaemonProcess API.

Specifically, allows `behaviorFactory` to return Option value. And removes generating process IDs by `numberOfProcess` and allows users to specify process IDs by `Set`.

**Example code**

```scala
val tags = Map(
    0 -> "tag-1",
    1 -> "tag-2",
    2 -> "tag-3",
)
val behaviorFactory: Int => Option[TagProcessor] =
    id => TagProcessor(tags.get(id))
val idSet = tags.keySet
ShardedDaemonProcess(system).init("TagProcessors", idSet, behaviorFactory)
```

`behaviorFactory` can indicate that it doesn't know a ID by `None` value. If ShardedDaemonProcess got `None` value, it simply ignores a request to start a process.

Users can remove any IDs by removing `Map` entries. The combination of IDs and tags does not change and it can be safely removed.

**Comment by johanandren** [comment]
I think for handling unknown it good enough to use `Behavior.stopped`, and that we do not need to introduce an `Option` there.

Having a set of ids is interesting though, I think that is worth exploring.

**Comment by patriknw** [comment]
> I think for handling unknown it good enough to use Behavior.stopped, and that we do not need to introduce an Option there.

I agree. Option not needed. We can also add some additional safety to catch any exceptions from the behaviorFactory, log, and treat that as stopped.

> specify process IDs by Set

I'm not sure I understand and I'm not convinced that is a good idea. Juggling two types of identifiers makes it more difficult to understand. Why can't it just be based on the index and the behaviorFactory can take it into account?

**Comment by negokaz** [comment]
@patriknw 

> Why can't it just be based on the index and the behaviorFactory can take it into account?

I considered the following situation.

**before a rolling update**
- **Node-1**: (id: 0, tag: "tag-1"), (id: 1, tag: "tag-2"), (id: 2, tag: "tag-3")
- **Node-2**: (id: 0, tag: "tag-1"), (id: 1, tag: "tag-2"), (id: 2, tag: "tag-3")
- **Node-3**: (id: 0, tag: "tag-1"), (id: 1, tag: "tag-2"), (id: 2, tag: "tag-3")

Remove "tag-1" from the application and update the nodes.

**updating Node-1**
- **Node-1**: (id: **0**, tag: "tag-2"), (id: **1**, tag: "tag-3")
- **Node-2**: (id: 0, tag: "tag-1"), (id: 1, tag: "tag-2"), (id: 2, tag: "tag-3")
- **Node-3**: (id: 0, tag: "tag-1"), (id: 1, tag: "tag-2"), (id: 2, tag: "tag-3")

In Node-1, "tag-2" and "tag-3" are identified by `id: 0` and `id: 1`.

If ClusterSharding assigned `id: 0` to Node-1 and `id: 1` to Node-2, "tag-2" behavior spawns in Node-1 and Node-2. It may lead data update conflicts.

To get around the problem, replace the "tag-1" behavior with `Behavior.stop` instead of removing.
It means that users must take care not to change the index and tag combination. But it's hard to notice.

ShardedDaemonProcess can solve this problem on its own, if it could detect differences from other nodes. For example, [KeepAlivePinger](https://github.com/akka/akka/blob/4947c07385e00fb99eb6fcbee55a1f64ede0fa13/akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessImpl.scala#L40) sends `StartEntity` along with `identities` and `ShardedDaemonProcess` compares its own `entityIds` and `identities` which was received from another node, and then if it is not the same, it ignores the `StartEntity`.

**Comment by patriknw** [comment]
I think I first need to understand the background for removing a tag? Is this eventsByTag based projections, by the way?

For example, if you remove tag-1 and start a rolling update you don't know if all events with tag-1 has been processed yet.

ShardedDaemonProcess is not designed for dynamically changing the id mapping. It's supposed to be a fixed number of processess. In general, a full cluster shutdown is needed without rolling update if the mapping is changed. That is similar for Cluster Sharding itself, changing number of shards or the extractor functions require full cluster shutdown.

For Sharding you would typically have more shards than number of nodes to support some dynamics. Typical recommendation is number of shards = 10x number of nodes. It's similar for ShardedDaemonProcess, you would use more tags than number of nodes.

We have indeed discussed ideas for supporting more dynamic scaling up and down, but it's a rather difficult topic that should also consider how to changing the tagging in EventSourcedBehavior in a dynamic way. For example splitting each tag into two, but still ensure that the processing order is not violated.

**Comment by negokaz** [comment]
> I think I first need to understand the background for removing a tag? Is this eventsByTag based projections, by the way? 

Yes, I assumed eventsByTag based projections.

> For example, if you remove tag-1 and start a rolling update you don't know if all events with tag-1 has been processed yet.

I believe I remove a tag when I no longer have interest in it. In other words, we don't have to wait for all events to be processed, I think.

Another scenario I had in mind was a rollback release.I would like to be able to undo the update with the new tags added.

> For Sharding you would typically have more shards than number of nodes to support some dynamics. Typical recommendation is number of shards = 10x number of nodes. It's similar for ShardedDaemonProcess, you would use more tags than number of nodes.

It is true that ClusterSharding does not support dynamic changes in the number of Shards. However, dynamic changes in the number of entities are supported. Is it difficult to consider a ShardedDaemonProcess as an entity?

I understand that it's difficult to perfect every rolling update scenario, but I think it's well worth it to make rolling updates work in some scenarios.
